xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -553,14 +555,17 @@ private void handleLdsResponseForConfigUpdate(DiscoveryResponse ldsResponse) {
     List<String> listenerNames = new ArrayList<>(ldsResponse.getResourcesCount());
     try {
       for (com.google.protobuf.Any res : ldsResponse.getResourcesList()) {
+        if (res.getTypeUrl().equals(ADS_TYPE_URL_LDS_v2)) {
+          res = res.toBuilder().setTypeUrl(ADS_TYPE_URL_LDS).build();
+        }
         Listener listener = res.unpack(Listener.class);
         listeners.add(listener);
         listenerNames.add(listener.getName());
       }
     } catch (InvalidProtocolBufferException e) {
       logger.log(XdsLogLevel.WARNING, "Failed to unpack Listeners in LDS response {0}", e);
       adsStream.sendNackRequest(
-          ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName),
+          ADS_TYPE_URL_LDS_v2, ImmutableList.of(ldsResourceName),
```

Comment:
My comment above about v2 vs v3. For now this is okay but when the v3 migration is complete and our request uses v3 typeUrl should we analyze the responses for v2 vs v3 typeUrl and use that for resource validation and also subsequent ACK/NACK?

Reply:
My understanding of the design doc is we should accept (ACK) both v2 and v3 response before we drop v2 support.

Reply:
Agreed but my point was slightly different. When the v3 migration is complete we will request v3 resources by using the v3 type Url in the DiscoveryRequest (https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L54). Let's say the control plane decides to return v2 resources instead (which is possible and that's why we need to maintain v2 support for some time). So in the DiscoveryResponse the typeUrl (https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L98) will be for v2 and so will be in the individual resources (https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L78). When that happens and you decide to ACK or NACK that response will that ACK/NACK use v2 typeUrl (to match the response) or the v3 typeUrl (to match the original request)? If the answer it to match the original request then there isn't much to do but if the answer is to match the response then we need additional logic (may be not in this PR).

Reply:
So in the case of (after migration, before dropping v2 support) xds client sending v3 request and receiving v2 response (and v2 individual resources of course): we parse the v2 resources to v3 resource objects and ACK the v2 response (with v2 [typeUrl](https://github.com/envoyproxy/envoy/blob/master/api/envoy/service/discovery/v3/discovery.proto#L98)  in the ACK I think. cc @markdroth to confirm) if no error occurs.

Reply:
Basically I think we should ACK/NACK with type_url version the same as the type_url of the response.

Reply:
Basically I think we should ACK/NACK with type_url version the same as the type_url of the response.

Reply:
In that case this line can be changed to 
```
ldsResponse.getTypeUrl(), ImmutableList.of(ldsResourceName),
```
and similar changes elsewhere for all ACK/NACK cases. Provided we confirm that the typeUrl in ACK/NACK has to match the one in the response. WDYT?

Reply:
Good question.  I think we will want the ACK/NACK to use the version that we originally requested, not the version that the server sent back.  This is because the xDS protocol does not have a completely separate ACK/NACK message; the ACK or NACK is implicitly included in the next request, which may also change the set of requested resource names at the same time as it sends an ACK/NACK.  And since we are already expecting the server to be understanding that the v2 and v3 resource names are the same, that logic should not be any different for ACK/NACK than it should be for any other request.

@htuch, can you give me a sanity check that the above is correct?

Reply:
@markdroth I can see it either way. If the server uses the typeUrl (from the response) for its bookkeeping and version tracking of resources for clients then ACKing/NACKing with the original request's typeUrl could confuse it. It will be good to get a definitive answer from the xDS spec. BTW, what does C-core do?

Reply:
In any case, using the typeUrl of the original request sounds more logical so let's go ahead with that for now. If we hear otherwise we can make that change as a separate PR.

Reply:
My work-in-progress PR for C-core (grpc/grpc#23281) will always use the version that the client is using, not the version that the server sends back.

The more I think about this, the less sense it makes to me to use the version that the server sent back.  Any server that decides to send a different version than the one that was asked for is already needing to understand the mapping between those two versions.  And having the client switch which version it's referring to in the middle of the stream seems like it could badly confuse the xDS semantics, because (as I said earlier) an ACK/NACK can be part of the same request as a change to the set of subscribed resource names, and the client will have already previously sent a request for a different set of resource names using the other version.

Reply:
I agree (with Mark's comment) and go ahead with that approach. Just to add for better/wider understanding, the server side should ignore the typeUrls in the ACK/NACK but keep using the typeUrl used to establish the stream (in my opinion).

Reply:
Yeah, ACK/NACK with the original version.